### ISSUE TITLE
Running inference on a live webcam feed

### ISSUE THREAD

**Issue by sagz** [PeterL1n/BackgroundMattingV2#3]
How would I go about running the inference model live on a webcam feed? (to be eventually piped into video conference apps etc).

Is there a way to do this as an OBS filter? And what hardware acceleration would be required to get enough performance? MacOS for example wouldn't be able to do CUDA really.

The end goal is to have this show up as a virtual camera that can be selected in whatever video ingestion platform you use

**Comment by PeterL1n** [comment]
For our research, we have written scripts to pipe live webcam input through our model and output to zoom as a virtual camera exactly like you said. Our script was only for Linux operating systems using PyTorch. We are working on publishing that as a demo to a separate repo.

As for running on Apple devices, we tried to export our model to CoreML. However, we found that CoreML currently is extremely slow for cropping out patches and replacing patches (an essential part of our model architecture). GPU utilization is very low. Either Apple needs to fix that or someone needs to write custom Metal code to speed it up. Currently, the CoreML export script is not published. If there are people in the community who are very experienced in this area and would like to help, I am open to sharing it.

Our model can also be exported to ONNX. ONNX has many inference backends for different device types but each has different levels of support and compatibility. Most backends have very good support for operations like convolutions since it is very commonly used, but cropping and replacing image patches are not optimized or not supported. If people in the industry have made it work well. Let us know.
